server/dbInit.ts

```diff
@@ -1,6 +1,156 @@
+    const subjectIfExists = await Subjects.findOne({ subShort: subject.value.toUpperCase() }).exec();
+
+    if (!subjectIfExists) {
+      console.log(`Adding new subject: ${subject.value}`);
+      const res = await new Subjects({
+        _id: shortid.generate(),
+        subShort: subject.value.toUpperCase(),
+        subFull: subject.descrformal,
+      }).save().catch((err) => null);
+
+      // db operation was not successful
+      if (!res) {
+        throw new Error();
+      }
+    }
+
+    return true;
+  })).catch((err) => null);
+
+  if (!v1) {
+    console.log("Something went wrong while updating subjects!");
+    return false;
+  }
+
+  // Update the Classes in the db
+  const v2 = await Promise.all(subjects.map(async (subject) => {
+    const classes = await fetchClassesForSubject(endpoint, semester, subject);
+
+    // skip if something went wrong fetching classes
+    // it could be that there are not classes here (in tests, corresponds to FEDN)
+    if (!classes) {
+      return true;
+    }
+
+    // Update or add all the classes to the collection
+    const v = await Promise.all(classes.map(async (cl) => {
+      const classIfExists = await Classes.findOne({ classSub: cl.subject.toUpperCase(), classNum: cl.catalogNbr }).exec();
+
+      // The class does not exist yet, so we add it
+      if (!classIfExists) {
+        console.log(`Adding new class ${cl.subject} ${cl.catalogNbr}`);
+        const res = await new Classes({
+          _id: shortid.generate(),
+          classSub: cl.subject.toUpperCase(),
+          classNum: cl.catalogNbr,
+          classTitle: cl.titleLong,
+          classFull: `${cl.subject.toUpperCase()} ${cl.catalogNbr} ${cl.titleLong}`,
+          classSems: [semester],
+          // TODO: provide initial values here?
+          classProfessors: null,
+          classRating: null,
+          classWorkload: null,
+          classDifficulty: null,
+        }).save().catch((err) => { console.log(err); return null; });
+
+        if (!res) {
+          console.log(`Unable to insert class ${cl.subject} ${cl.catalogNbr}!`);
+          throw new Error();
+        }
+      } else { // The class does exist, so we update semester information
+        console.log(`Updating class information for ${classIfExists.classSub} ${classIfExists.classNum}`);
+
+        // Compute the new set of semesters for this class
+        const classSems = classIfExists.classSems.indexOf(semester) == -1 ? classIfExists.classSems.concat([semester]) : classIfExists.classSems;
+
+        // update db with new semester information
+        const res = await Classes.findOneAndUpdate({ _id: classIfExists._id }, { $set: { classSems } }).exec()
+          .catch((err) => null);
+
+        if (!res) {
+          console.log(`Unable to update class information for ${cl.subject} ${cl.catalogNbr}!`);
+          throw new Error();
+        }
+      }
+
+      return true;
+    })).catch((err) => null);
```

Comment:
Same as above, print this error

Reply:
This only fails in the case where something else has already failed, in which case we will have gotten another, more detailed error message ahead of time.

Reply:
Makes sense